apache/metron
# METRON-1661: Create Pcap Query Filter endpoint

## Contributor Comments
This PR adds a REST endpoint for submitting a query filter type Pcap query request.

### Changed Included
- A new Pcap query request endpoint that submits a query filter query
- Slight refactor of the PcapService and PcapServiceImpl class that changes the fixed method to a submit method that can handle either a fixed pcap request of query pcap request
- Unit and integration tests

### Testing

This has been tested in full dev.  Testing instructions are as follows:

1. Spin up full dev
2. Create the HDFS directories specified in application.yml
  - /apps/metron/pcap/input
  - /apps/metron/pcap/interim
  - /apps/metron/pcap/output
3. Put pcap data in `/apps/metron/pcap/input`
4. Run a query pcap query to generate results (note the query property):
```
curl -X POST --header 'Content-Type: application/json' --header 'Accept: application/json' -d '{
  "endTimeMs": 1458240269424,
  "startTimeMs": 1458240269419,
  "query":"ip_dst_port == 22"
}' 'http://node1:8082/api/v1/pcap/query'
```
5. Monitor the job with the get status endpoint until the job has SUCCEEDED and returns a pageTotal:
```
{
  "jobId": "job_1532012905403_0001",
  "jobStatus": "SUCCEEDED",
  "description": "map: 20.0%, reduce: 0.0%",
  "percentComplete": 100,
  "pageTotal": 1
}
```
Once the job has finished results should be present in HDFS in a location that is similar to `/apps/metron/pcap/output/user/MAP_REDUCE/job_1532012905403_0001/page-1.pcap` (job id will be different).  The pcaps in the results should match the condition specified in the "query" property in the request.

## Pull Request Checklist

Thank you for submitting a contribution to Apache Metron.  
Please refer to our [Development Guidelines](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=61332235) for the complete guide to follow for contributions.  
Please refer also to our [Build Verification Guidelines](https://cwiki.apache.org/confluence/display/METRON/Verifying+Builds?show-miniview) for complete smoke testing guides.  


In order to streamline the review of the contribution we ask you follow these guidelines and ask you to double check the following:

### For all changes:


### For code changes:
  ```
  mvn -q clean integration-test install && dev-utilities/build-utils/verify_licenses.sh 
  ```


### For documentation related changes:

  ```
  cd site-book
  mvn site
  ```

#### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.
It is also recommended that [travis-ci](https://travis-ci.org) is set up for your personal repository such that your branches are built there before submitting a pull request.
